### PR TITLE
lib: Add attrsets.mainProgram

### PR DESCRIPTION
**Pull request by nicoonoclaste**
###### Motivation for this change

Writing `"${pkgs.something}/bin/something"` is a common idiom, both in `nixpkgs` itself and Nix users' configuration.
That idiom is brittle, as it will break if the binary is renamed ~~or the package is made multi-output~~.

Using instead `lib.mainProgram pkgs.something` avoids both of those issues, and avoids implementing ad-hoc solutions
everywhere, while often being shorter.

As an example, `node-packages` is simplified and made more uniform; prior to the change, `PUPEETER_EXECUTABLE_PATH`
was defined in several different — but equivalent — ways using `.outPath` or not; using `lib.mainProgram` is 
equivalent to both of those and less brittle.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- ~~For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))~~
- ~~Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))~~
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- ~~Tested execution of all binary files (usually in `./result/bin/`)~~
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - ~~(Package updates) Added a release notes entry if the change is major or breaking~~
  - ~~(Module updates) Added a release notes entry if the change is significant~~
  - ~~(Module addition) Added a release notes entry if adding a new NixOS module~~
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by alyssais** [comment]
> That idiom is brittle, as it will break if the binary is renamed or the package is made multi-output.

Small correction: conventionally we put binaries in the default output, specifically so this doesn't break.  I'm not aware of many cases of binaries being renamed while still being compatible either, fwiw.

**Comment by nbraud** [comment]
> > That idiom is brittle, as it will break if the binary is renamed or the package is made multi-output.
> 
> Small correction: conventionally we put binaries in the default output, specifically so this doesn't break. I'm not aware of many cases of binaries being renamed while still being compatible either, fwiw.

Thanks for the correction, I amended the PR description.

**Comment by alyssais** [comment]
<pre><samp>nix-repl> lib.mainProgram bash
error: attribute 'mainProgram' missing, at /home/src/nixpkgs/lib/attrsets.nix:500:43</samp></pre>

I think for this to be useful, it would have to work the same [way `nix run` works in unstable Nix](https://nixos.org/manual/nix/unstable/command-ref/new-cli/nix3-run.html#description), where it falls back when mainProgram isn't defined.

**Comment by nbraud** [comment]
> nix-repl> lib.mainProgram bash
> error: attribute 'mainProgram' missing, at /home/src/nixpkgs/lib/attrsets.nix:500:43
> 
> I think for this to be useful, it would have to work the same [way `nix run` works in unstable Nix](https://nixos.org/manual/nix/unstable/command-ref/new-cli/nix3-run.html#description), where it falls back when mainProgram isn't defined.

OK; I was thinking of PRing definitions of `mainProgram` where useful, but I guess falling back to the eponymous program is a smarter move.

Is there a way to check whether the file exists and is executable, so I can provide a nicer error message in that case?

**Comment by alyssais** [comment]
> OK; I was thinking of PRing definitions of `mainProgram` where useful, but I guess falling back to the eponymous program is a smarter move.

That's how I'd have preferred it would work, tbh, but that ship seems to have sailed — it wouldn't make sense to have this function work differently from how Nix does it.

> Is there a way to check whether the file exists and is executable, so I can provide a nicer error message in that case?

Not in lib, because the package hasn't necessarily been build at eval time.

**Comment by nbraud** [comment]
> > OK; I was thinking of PRing definitions of `mainProgram` where useful, but I guess falling back to the eponymous program is a smarter move.
> 
> That's how I'd have preferred it would work, tbh, but that ship seems to have sailed — it wouldn't make sense to have this function work differently from how Nix does it.

Same, I just didn't remember in the moment that `nix run` is a thing.

> > Is there a way to check whether the file exists and is executable, so I can provide a nicer error message in that case?
> 
> Not in lib, because the package hasn't necessarily been build at eval time.

That's what I guessed, but I'm not super-familiar w/ some language details so I figured I'd rather ask.  Thanks for confirming  :purple_heart:

**Comment by figsoda** [comment]
should this be renamed to `getMainProgram` instead?

File: lib/attrsets.nix [hunk 1]
```diff
@@ -496,6 +496,10 @@ rec {
   getDev = getOutput "dev";
   getMan = getOutput "man";
 
+  /* Get a package's main executable's path */
```
Comment (samuelgrf) [inline lib/attrsets.nix:499]:
```suggestion
  /* Get path to main executable of a package */
```

**Comment by alyssais** [comment]
> should this be renamed to `getMainProgram` instead?

`mainProgram` is pretty clear to me — is there something else you'd expect a function with that name to do?

**Comment by nbraud** [comment]
> should this be renamed to `getMainProgram` instead?

As Alyssa said, that seemed unambiguous and clear, while still being concise.
I could change it if it is ambiguous to others or inconsistent with other lib functions, though.

File: lib/attrsets.nix (hunk 1 above)
Comment (nbraud) [reply lib/attrsets.nix]: @samuelgrf The suggested change seems like unidiomatic English to me, but I just pushed an alternative formulation without nested `'s`.  Can you confirm this solves the issue you saw, or clarify what was the issue?

**Comment by figsoda** [comment]
I feel like `getMainProgram` is more consistent with the existing `getName`, `getVersion`, and other `get*`s
`mainProgram` is not very ambiguous, but there are `lib.maintainers` and `lib.platforms` which are also meta attributes

**Comment by nbraud** [comment]
> `mainProgram` is not very ambiguous, but there are `lib.maintainers` and `lib.platforms` which are also meta attributes

I don't understand what you mean here, this seems like a nonsequitur to me.
How do `lib.{maintainers, platforms}` (which are attrsets) relate to the proposed function?

File: lib/attrsets.nix (hunk 1 above)
Comment (samuelgrf) [reply lib/attrsets.nix]: I guess something about the nested `'s` just didn't sound right to me.
This seems more like a personal preference now that I think about it, you can change it back if you want.

The alternative formulation sounds more natural than my suggestion, so I'll mark this as resolved.

Thanks for your time!

**Comment by figsoda** [comment]
I think `mainProgram` is fine, I just personally prefer `getMainProgram`

**Comment by nbraud** [comment]
> I think `mainProgram` is fine, I just personally prefer `getMainProgram`

'k, thanks for clarifying  :)

File: lib/attrsets.nix (hunk 1 above)
Comment (nbraud) [reply lib/attrsets.nix]: You are welcome, and thanks for pointing out the comment didn't sound very natural  :)

**Comment by roberth** [comment]
I'd prefer `mainProgram` to be a recursively defined `passthru` attribute that already includes the correct output path. Requires #119942. It'd look like
```
passthru.mainProgram = "${self}/bin/main";
```
and
```nix
runCommand "greeting-text" { } ''
  ${hello.mainProgram} >$out
''
```

No function required.

Also `nativeBuildInputs` exists and should be used because it makes cross compilation work. Of course that doesn't always apply, but it is noteworthy.

**Comment by roberth** [comment]
I guess `mainProgram` could be considered "taken" with the specific meaning that it already has. `exe` would be a short and sweet alternative. I find `"${hello.exe} >$out"` oddly satisfying.

**Comment by nbraud** [comment]
> I'd prefer `mainProgram` to be a recursively defined `passthru` attribute that already includes the correct output path. Requires #119942. It'd look like [...]
> ```nix
> runCommand "greeting-text" { } ''
>   ${hello.mainProgram} >$out
> ''
> ```

That's an interesting approach, but #119942 has received no attention from reviewers since May, and currently has merge conflicts, so it seems inappropriate to block other changes on it.

**Comment by nbraud** [comment]
Rebased to deal with merge conflict

**Comment by nbraud** [comment]
@zimbatm I believe Alyssa would like your opinion prior to merging this.

**Comment by Infinisil** [comment]
> I'd prefer `mainProgram` to be a recursively defined `passthru` attribute that already includes the correct output path. Requires #119942. It'd look like
> 
> ```
> passthru.mainProgram = "${self}/bin/main";
> ```
> 
> and
> 
> ```nix
> runCommand "greeting-text" { } ''
>   ${hello.mainProgram} >$out
> ''
> ```
> 
> No function required.

I really like this, but we don't even need it to be this verbose. Support for this can be embedded into the `mkDerivation` function itself, such that it automatically sets `passthru.mainProgram = "${self}/bin/${meta.mainProgram}"`

Since then there would be both `bash.mainProgram` and `bash.meta.mainProgram`, we should probably rename one of them so it's not confusing. E.g. we could call the latter `meta.mainProgramName` instead.

**Comment by Infinisil** [comment]
Oh, I didn't realize `meta.mainProgram` is already a thing, never mind the renaming then

**Comment by Infinisil** [comment]
Something like this seems to work well:
```diff
diff --git a/pkgs/stdenv/generic/make-derivation.nix b/pkgs/stdenv/generic/make-derivation.nix
index 56cfa0c503f..01a474ba26c 100644
--- a/pkgs/stdenv/generic/make-derivation.nix
+++ b/pkgs/stdenv/generic/make-derivation.nix
@@ -370,9 +370,8 @@ else let
                    else true);
     };
 
-in
 
-lib.extendDerivation
+self = lib.extendDerivation
   validity.handled
   ({
      overrideAttrs = f: stdenv.mkDerivation (attrs // (f attrs));
@@ -402,10 +401,13 @@ lib.extendDerivation
        args = [ "-c" "export > $out" ];
      });
 
+     mainProgram = "${lib.getBin self.outPath}/bin/${meta.mainProgram or derivationArg.pname or (builtins.parseDrvName derivationArg.name).name}";
+
      inherit meta passthru;
    } //
    # Pass through extra attributes that are not inputs, but
    # should be made available to Nix expressions using the
    # derivation (e.g., in assertions).
    passthru)
-  (derivation derivationArg)
+  (derivation derivationArg);
+ in self
```

File: lib/attrsets.nix [hunk 2]
```diff
@@ -496,6 +496,10 @@ rec {
   getDev = getOutput "dev";
   getMan = getOutput "man";
 
+  /* Get the path to a package's main executable */
+  mainProgram = pkg:
```
Comment (zimbatm) [inline lib/attrsets.nix:500]: What do you think `mainBin` to make it a bit shorter? I know it's bike-shedding territory, feel free to ignore.

**Review by zimbatm — APPROVED** [review]
Looks good to me. nixpkgs can benefit a lot from having a utility like that.

File: lib/attrsets.nix (hunk 2 above)
Comment (roberth) [reply lib/attrsets.nix:500]: Please let's keep `bin` as the directory. The executable files can be `exe` if we want to keep it short.

**Comment by roberth** [comment]
> That's an interesting approach, but #119942 has received no attention from reviewers

Then why not review it?

> since May, and currently has merge conflicts, so it seems inappropriate to block other changes on it.

Why would that be?

------

I think we should ask ourselves the question whether we want Nixpkgs to have a good design or just pile convenience upon convenience until we have a big pile of unnecessary code.

**Review by roberth — CHANGES_REQUESTED** [review]
This should be solved with an attribute, not a function.

**Comment by roberth** [comment]
Actually you don't have to wait for #119942 because you can implement the logic for this directly in `mkDerivation`, based on `meta.mainProgram`.

**Comment by adrian-gierakowski** [comment]
> I'd prefer `mainProgram` to be a recursively defined `passthru` attribute that already includes the correct output path. Requires #119942. It'd look like
> 
> ```
> passthru.mainProgram = "${self}/bin/main";
> ```
> 
> and
> 
> ```nix
> runCommand "greeting-text" { } ''
>   ${hello.mainProgram} >$out
> ''
> ```
> 
> No function required.
> 
> Also `nativeBuildInputs` exists and should be used because it makes cross compilation work. Of course that doesn't always apply, but it is noteworthy.

I've been using the same approach via a helper function which takes a derivation and adds `passthru.defaultBinPath` to it. I'd love to see this become a standard in nixpkgs

**Comment by roberth** [comment]
> I don't think it's the best way to do it, but here you go.

This is: https://github.com/NixOS/nixpkgs/pull/158461

**Comment by Infinisil** [comment]
Being worried about the performance of adding an attribute to each derivation makes me wonder about the optimization potential by removing existing attributes:

```
nix-repl> lib.attrNames hello 
[ "__ignoreNulls" "all" "args" "buildInputs" "builder" "configureFlags" "depsBuildBuild" "depsBuildBuildPropagated" "depsBuildTarget" "depsBuildTargetPropagated" "depsHostHost" "depsHostHostPropagated" "depsTargetTarget" "depsTargetTargetPropagated" "doCheck" "doInstallCheck" "drvAttrs" "drvPath" "inputDerivation" "meta" "name" "nativeBuildInputs" "out" "outPath" "outputName" "outputs" "override" "overrideAttrs" "overrideDerivation" "passthru" "patches" "pname" "propagatedBuildInputs" "propagatedNativeBuildInputs" "src" "stdenv" "strictDeps" "system" "tests" "type" "userHook" "version" ]
```

Many of those probably aren't really needed. Note that these are added by [the `derivation` primop](https://github.com/NixOS/nix/blob/2.7.0/src/libexpr/primops/derivation.nix) itself, but we could theoretically sidestep that by using the `derivationStrict` primop directly.

**Comment by roberth** [comment]
> Many of those probably aren't really needed.

Those guts should never have been exposed.
Also we should program with proper composition and less `removeAttrs args`. I suspect that will save a lot too. It's the kind of "radical" change we need for [`outputOf`](https://github.com/NixOS/rfcs/blob/master/rfcs/0092-plan-dynamism.md) anyway.

Maybe create an issue if you want to discuss this more? It's a bit off topic here.

**Comment by Artturin** [comment]
merged https://github.com/NixOS/nixpkgs/pull/167247

**Comment by roberth** [comment]
Thank you @nbraud for starting the conversation and doing the work, even if another implementation got merged in the end.

**Pull request closed without merging**